### PR TITLE
Create account

### PR DESCRIPTION
**Pull request by alishamohanty**

### CODE REVIEW

File: resources/migrations/20211215052838-create-user.up.sql [hunk 1]
```diff
@@ -0,0 +1,8 @@
+CREATE TABLE users(
+    id UUID PRIMARY KEY DEFAULT uuid_generate_v4(),
+    name TEXT NOT NULL,
+    role TEXT NOT NULL,
+    email TEXT NOT NULL UNIQUE,
+    password TEXT NOT NULL
```
Comment (govindkrjoshi) [inline resources/migrations/20211215052838-create-user.up.sql:6]: Please add a `created_at` and `updated_at` to the table.

File: src/winter_onboarding_2021/fleet_management_service/handlers/user.clj [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+(defn signup-form [_]
+  {:title "Sign-up"
+   :content (view/signup-form)})
+
+(defn user-exist? [email]
+  (not-empty (user-model/find-by-keys {:email email})))
+
+
+(defn create-user [{:keys [multipart-params]}]
```
Comment (govindkrjoshi) [inline src/winter_onboarding_2021/fleet_management_service/handlers/user.clj:26]: Why are you using `multipart-params` here?

File: src/winter_onboarding_2021/fleet_management_service/handlers/user.clj [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  (let [validated-user (s/conform ::specs/signup-form multipart-params)
+        user-exist (user-exist? (:email multipart-params))]
+    (if (or (s/invalid? validated-user) user-exist)
+      (if user-exist
+        (-> (flash-msg "User already exisits, use different email!" false)
```
Comment (govindkrjoshi) [inline src/winter_onboarding_2021/fleet_management_service/handlers/user.clj:31]: Since it's only one function call, it might be more readable to avoid threading.

File: src/winter_onboarding_2021/fleet_management_service/handlers/user.clj [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+            (merge (response/redirect "/users/signup")))
+        (-> (flash-msg "Could not create user, enter valid details!" false)
+            (merge (response/redirect "/users/signup"))))
+      (let [_ (user-model/create (assoc validated-user
```
Comment (govindkrjoshi) [inline src/winter_onboarding_2021/fleet_management_service/handlers/user.clj:35]: Why do you need the let block if you're not binding the value?

File: src/winter_onboarding_2021/fleet_management_service/handlers/user.clj [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+                                        :role "admin"
```
Comment (govindkrjoshi) [inline src/winter_onboarding_2021/fleet_management_service/handlers/user.clj:36]: Why are all users created as an `admin`?

File: src/winter_onboarding_2021/fleet_management_service/specs.clj [hunk 6]
```diff
@@ -24,3 +24,20 @@
   (s/keys :req [:cabs/name
                 :cabs/licence-plate
                 :cabs/distance-travelled]))
+
+;; Users
+(s/def :users/name string?)
```
Comment (govindkrjoshi) [inline src/winter_onboarding_2021/fleet_management_service/specs.clj:29]: [Naming] Do you think it's better to use singular for the spec names? For example: `:user/name` since the name is on a single user, and not a group of users. 

In the same spirit, naming the `users` spec to `user`

File: src/winter_onboarding_2021/fleet_management_service/views/user.clj [hunk 7]
```diff
@@ -0,0 +1,15 @@
+(ns winter-onboarding-2021.fleet-management-service.views.user)
+
+
+(defn signup-form []
+  [:form {:action "/users/signup" :method "POST" :enctype "multipart/form-data"}
+   [:div {:class "mb-3" :id "email"}
```
Comment (govindkrjoshi) [inline src/winter_onboarding_2021/fleet_management_service/views/user.clj:6]: We should try to extract smaller components and reuse them where we can. I think here we can extract the input and label in to a component. 

It should also be possible to also make the form itself a component which takes a form spec and renders it.

File: test/winter_onboarding_2021/fleet_management/db/user_test.clj [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+(use-fixtures :once fixtures/config fixtures/db-connection)
+(use-fixtures :each fixtures/clear-db)
+
+(deftest create-user
+  (testing "Should create a user in the users table"
+    (let [user {:name "Harry Potter"
+                :role "admin"
+                :email "harry@hogwarts.edu"
+                :password "hermione@123"}
+          _ (user-db/create user)]
```
Comment (govindkrjoshi) [inline test/winter_onboarding_2021/fleet_management/db/user_test.clj:15]: Why not just directly pass the map to `user-db/create`?

Comment (govindkrjoshi) [reply test/winter_onboarding_2021/fleet_management/db/user_test.clj:15]: Please add more test cases here? What if email is nil, or name is nil? Since this is the DB namespace it should test all the edge cases thoroughly.

File: test/winter_onboarding_2021/fleet_management/factories.clj [hunk 9]
```diff
@@ -10,3 +10,6 @@
 
 (defn create-cabs [num]
   (gen/generate (generate-cab num)))
+
+(defn generate-user []
```
Comment (govindkrjoshi) [inline test/winter_onboarding_2021/fleet_management/factories.clj:14]: `generate-signup-params` ?

File: test/winter_onboarding_2021/fleet_management/fixtures.clj [hunk 10]
```diff
@@ -23,6 +23,7 @@
 
 (defn clear-db [f]
   (jdbc/execute! db/db-conn  ["TRUNCATE TABLE cabs"])
+  (jdbc/execute! db/db-conn  ["TRUNCATE TABLE users"])
```
Comment (govindkrjoshi) [inline test/winter_onboarding_2021/fleet_management/fixtures.clj:26]: truncate can take multiple table names.

Comment (govindkrjoshi) [reply test/winter_onboarding_2021/fleet_management/fixtures.clj:26]: You should also look at using truncate with cascade.

File: test/winter_onboarding_2021/fleet_management/views/user_test.clj [hunk 11]
```diff
@@ -0,0 +1,15 @@
+(ns winter-onboarding-2021.fleet-management.views.user-test
+  (:require [clojure.test :refer [deftest is testing]]
+            [hiccup-find.core :as hf]
+            [winter-onboarding-2021.fleet-management-service.views.user :as view]))
+
+(deftest signup-form
```
Comment (govindkrjoshi) [inline test/winter_onboarding_2021/fleet_management/views/user_test.clj:6]: Also test for the action that the form makes.

**Review by govindkrjoshi — CHANGES_REQUESTED** [review]

File: src/winter_onboarding_2021/fleet_management_service/handlers/user.clj (hunk 4 above)
Comment (shhivam) [reply src/winter_onboarding_2021/fleet_management_service/handlers/user.clj:35]: We are now customising the successful account creation flash message to include the user's name & that's why we need to use `let`

File: src/winter_onboarding_2021/fleet_management_service/handlers/user.clj (hunk 5 above)
Comment (shhivam) [reply src/winter_onboarding_2021/fleet_management_service/handlers/user.clj:36]: For now, the default role is `admin` & as we add more roles in our codebase, we will modify this code accordingly.

File: src/winter_onboarding_2021/fleet_management_service/views/user.clj (hunk 7 above)
Comment (shhivam) [reply src/winter_onboarding_2021/fleet_management_service/views/user.clj:6]: >It should also be possible to also make the form itself a component which takes a form spec and renders it.

What exactly do you mean by form spec? Will it be used to pre-fill the form for editing? or do you want to make a general form component?

cc: @yogirajh007

File: src/winter_onboarding_2021/fleet_management_service/specs.clj (hunk 6 above)
Comment (yogirajh007) [reply src/winter_onboarding_2021/fleet_management_service/specs.clj:29]: We named it that way because of the table names in db. JDBC namespaces the keys by the tablename.
